sub-packages/bionemo-geometric/src/bionemo/geometric/base_featurizer.py

```diff
@@ -0,0 +1,105 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-Apache2
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from abc import ABC, abstractmethod, abstractproperty
+from typing import List
+
+from rdkit.Chem import Atom, Mol
+
+
+class BaseFeaturizer(ABC):
```

Comment:
I'd vote for a `Protocol` definition rather than requiring a strict inheritance pattern here:

https://typing.readthedocs.io/en/latest/spec/protocol.html#protocols

Probably worth having separate protocols for atom, bond, and molecule features

Reply:
I can make them `Protocol`. How does this look?

`
class BaseAtomFeaturizer(Protocol):
    """Abstract base featurizer class for all atom featurizer classes."""

    @property
    def n_dim(self) -> int:
        """Number of dimensions of computed feature."""
        ...

    def get_atom_features(self):
        """Function for getting features."""
        ...

    def __call__(self):
        """Call featurizer on atom"""
        ...
class AtomicNumberFeaturizer:
    """Class for featurizing atom by its atomic number."""

    def __init__(self, max_atomic_num: Optional[int] = None) -> None:
        """Initializes AtomicNumberFeaturizer class."""
        self.max_atomic_num = max_atomic_num if max_atomic_num else MAX_ATOMIC_NUM

    def n_dim(self) -> int:
        """Returns dimensionality of the computed features."""
        return self.max_atomic_num

    def get_atom_features(self, mol: Mol, atom_indices: Optional[Iterable] = None) -> List[int]:
        """Computes features of atoms of all of select atoms."""
        _atom_indices = atom_indices if atom_indices else range(mol.GetNumAtoms())
        return [mol.GetAtomWithIdx(a).GetAtomicNum() for a in _atom_indices]

    
    def __call__(self, atom, atom_indices: Optional[Iterable] = None) -> List[int]:
        return self.get_atom_features(atom)`